### DISCUSSION TITLE
reduce reliance on containers for increased code portability

### DISCUSSION THREAD

**Discussion by jahess** [openthread/openthread#8598] [Ideas]
I suggest developer reliance on containers results in software portability issues.  This is ironic because containers are designed to help with dependency issues.

But, in a developer context, containers mask portability issues.  It is too easy to avoid actually resolving portability issues because the software "works" in... the approved container.

I know this runs counter to prevailing practice.  But it seems inefficient to having to use multiple images / containers to do development.  For an example, see:

    https://devzone.nordicsemi.com/f/nordic-q-a/95270/please-modernize-to-the-latest-raspios-openthread-otbr-matter

So... build on multiple different platforms.  Use explicitly sized types.  Judiciously use wrappers or #ifs for platforms and library version dependencies.  But, hide these details from the clients of your code.

Thanks for consideration,
Jonathan Hess

**Comment by jwhui** [comment]
@jahess , thanks for providing your thoughts.

> Build on multiple different platforms.

We leverage GitHub Actions to implement build checks across a variety of platforms including different architectures, toolchains, and OSes - see [.github/workflows/build.yml](https://github.com/openthread/openthread/blob/main/.github/workflows/build.yml) as an example.

> Use explicitly sized types. Judiciously use wrappers or #ifs for platforms and library version dependencies. But, hide these details from the clients of your code.

The [STYLE_GUIDE.md](https://github.com/openthread/openthread/blob/main/STYLE_GUIDE.md) includes the following text:

- _Standard, scalar data types defined in stdint.h (C) or cstdint (C++) should be used for basic signed and unsigned integer types, especially when size and serialization to non-volatile storage or across a network is concerned. Examples of these are: uint8_t, int8_t, etc._

> Judiciously use wrappers or #ifs for platforms and library version dependencies

OpenThread defines a [platform abstraction](https://github.com/openthread/openthread/tree/main/include/openthread/platform) to minimize platform-specific dependencies within core code.

Of course, there is always room to improve. We welcome any contributions that help improve on these areas. Thanks!
